### ISSUE TITLE
Configurable CSS image inlining

### ISSUE THREAD

**Issue by kemsky** [angular/angular-cli#21569]
# 🚀 Feature request

### Command (mark with an `x`)

- [x] build
- [x] serve

### Description

Image inlining is nothing new and despite HTTP/2 arguments it still makes sense to inline small images to simplify deployment and optimize app loading and experience (e.g. no icon flickering). 

Previously it was working out of the box, then the feature was removed, now with `PostCSS` loader and `postcss-url` plugin it is very easy to add this feature back.

### Describe the solution you'd like

`postcss-url` provides rich CSS assets manipulation options e.g. filter by type, size, path, etc.

### Describe alternatives you've considered

The only alternative I was able to find is very ugly, it is webpack config patch for custom-webpack:

```ts
const transformer = PostCssUrlPlugin([
        {
            filter: '**/*.svg',
            url: 'inline'
        },
        {
            filter: '**/*.png',
            url: 'inline'
        }
    ]);

    let patch: any = null;

    let success = false;

    for (const rule of config.module.rules) {
        if (typeof rule !== 'string' && Array.isArray(rule.rules)) {
            for (const loader of rule.rules) {
                if (typeof loader === 'string') {
                    continue;
                }

                if (Array.isArray(loader.oneOf)) {
                    for (const oneOf of loader.oneOf) {
                        if (Array.isArray(oneOf.use)) {
                            for (const use of oneOf.use) {
                                if (typeof use === 'object' && use.loader.includes('postcss-loader')) {
                                    const options: any = use.options;
                                    const postcssOptions = options.postcssOptions;

                                    // https://github.com/angular/angular-cli/blob/e95ecb8ab0382eb803741619c446d6cc7b215ba0/packages/angular_devkit/build_angular/src/webpack/configs/styles.ts#L164
                                    if (typeof postcssOptions === 'function' && postcssOptions !== patch) {
                                        patch = (...args: any[]) => {
                                            const opts = postcssOptions.call(null, ...args);
                                            const has = opts.plugins.find((x: any) => x === transformer) != null;
                                            if (!has) {
                                                opts.plugins.push(transformer);
                                            }
                                            return opts;
                                        };

                                        options.postcssOptions = patch;

                                        success = true;
                                    }
                                }
                            }
                        }
                    }
                }
            }
        }
    }
```

**Comment by alan-agius4** [comment]
We did used to have this feature, but it was removed a couple of major versions ago, because in most cases inlining of assets is a performance foot-gun. 

For more information see https://github.com/angular/angular-cli/pull/11826#issuecomment-416260575 were the Chrome team provided a number of reasons why inlining of assets is discouraged which eventually lead to removing the feature from the CLI altogether.

**Comment by kemsky** [comment]
@alan-agius4, we create and support enterprise applications that are used primarily as desktop apps e.g. CRMs so smooth experience is our goal.

> @addyosmani
> Thanks for reaching out about when to inline resources. It was unclear from the thread whether you're distinguishing how you handle different types (web fonts, images etc).
> 
> We typically don't see a large number of sites inlining web fonts. Given 60% of origins load at least one web font and occasionally it's multiple families, I'm not sure that we should encourage breaking caching by inlining fonts.
> 
> A more popular pattern is preloading your fonts or images if they're critical to early paints to ensure the browser fetches them earlier but without breaking caching. Inlining images (e.g. base-64ing) can have negative perf connotations on mobile devices (last I heard).
> 
> Supporting easier inlining seems okay as long as it's an opt-in config authors use themselves and there's documentation warning them about the caveats.

Caching is not a big deal because typically js bundle size is much bigger  (100 times if you compare to the size of SVG icon) than the size of the single resource image/font. Browser spends time loading small files, often you see icons slowly appearing here and there. It is getting worse when the user is offline - UI became a mess, so you naturally want to preload resources when possible (that's what angular service worker does).

> @KenjiBaheux:
> 
> Inlining:
> 
> is bad for repeat visits
> is bad for performance (e.g. delays parsing of the rest of the resource where the inlining is happening)
> can lead to performance footguns (i.e. forgotten resources that are no longer needed, or inlining the whole family when only one face is needed / critical)
> makes font-display pointless (the font has been loaded regardless in all cases)
> bloats the bytes count
> seems like it would be fast for the UA to consume but it actually isn't the case (still has round trips and main thread business interruptions happening <= I think we tried to fix some of that but unclear if we succeeded, in part we also don't want to encourage inlining).
> My take on this is that we shouldn't do inlining

JS bundle is cached, so repeat visits are totally OK. Delayed parsing can be controlled using lazy modules. Other points are simply not relevant.

**Comment by alan-agius4** [comment]
Hi @kemsky,

Thanks for this feature request. We did discuss this during one of our meetings and we came to the conclusion that we shouldn't add this feature as part of the Angular CLI. The main reasons for this are;

- Typically you'd only want to embed a limited number of assets under a certain size in the critical path and is no easy way to do this automatically.
- This feature would require a  number of options such as for: allowed file paths, allowed file formats, max size.
- [Preloading hints](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) are to be preferred.
- Encoding an image to base64 manually and adding to the stylesheet is trivial.
- Users wishing to embed assets during the build can use 3rd party builders, where they can configure and fine grain their assets embedding rules.
- Base64 data is bigger than their original binary representations up to 33%.
- Web browser need to wait for the whole base64 stream before continuing to render the CSS (even when the image is not displayed. This would impact negatively on your Core Web Vital metrics.
- Extensive documentation would be need as feature comes with a number of caveats and pitfalls.

Overall, if you are experience issues that images are taking too long to appear on screen [preloading hints](https://developer.mozilla.org/en-US/docs/Web/HTML/Link_types/preload) should be preferred.
